File: backend/src/main/java/com/codeforge/module/problem/controller/ProblemController.java
package com.codeforge.module.problem.controller;

import com.codeforge.common.dto.ApiResponse;
import com.codeforge.common.dto.PagedResponse;
import com.codeforge.common.enums.ProblemDifficulty;
import com.codeforge.module.problem.dto.response.ProblemDetailResponse;
import com.codeforge.module.problem.dto.response.ProblemSummaryResponse;
import com.codeforge.module.problem.service.ProblemService;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/v1/problems")
@RequiredArgsConstructor
public class ProblemController {

    private final ProblemService problemService;

    @GetMapping
    public ResponseEntity<ApiResponse<PagedResponse<ProblemSummaryResponse>>> getProblems(
            @RequestParam(required = false) ProblemDifficulty difficulty,
            @RequestParam(required = false) String search,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size
    ) {
        Pageable pageable = PageRequest.of(page, Math.min(size, 50), Sort.by("id").ascending());
        PagedResponse<ProblemSummaryResponse> response = problemService.getProblems(difficulty, search, pageable);
        return ResponseEntity.ok(ApiResponse.success(response, "Problems fetched successfully"));
    }

    @GetMapping("/{slug}")
    public ResponseEntity<ApiResponse<ProblemDetailResponse>> getProblemBySlug(
            @PathVariable String slug
    ) {
        ProblemDetailResponse response = problemService.getProblemBySlug(slug);
        return ResponseEntity.ok(ApiResponse.success(response, "Problem fetched successfully"));
    }
}

File: backend/src/modules/problem/problem.controller.js
import { ProblemService } from './problem.service.js';
import { sendApiResponse } from '../../common/utils/apiHelpers.js';

export class ProblemController {
  static async getProblems(req, res, next) {
    try {
      const difficulty = req.query.difficulty;
      const search = req.query.search;
      const page = parseInt(req.query.page || '0', 10);
      const size = parseInt(req.query.size || '10', 10);

      const result = await ProblemService.getProblems(difficulty, search, page, size);
      return sendApiResponse(res, result, 'Problems fetched successfully');
    } catch (err) {
      next(err);
    }
  }

  static async getProblemBySlug(req, res, next) {
    try {
      const { slug } = req.params;
      const result = await ProblemService.getProblemBySlug(slug);
      return sendApiResponse(res, result, 'Problem fetched successfully');
    } catch (err) {
      next(err);
    }
  }
}
